### ISSUE TITLE
_SIGCHLDWaker in twisted.internet.posixbase doesn't check its pipe is still open when registering signal handlers

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5151]
|<img alt="hodgestar's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| hodgestar reported|
|-|-|
|Trac ID|trac#5151|
|Type|defect|
|Created|2011-06-10 12:11:42Z|
|Branch|https://github.com/twisted/twisted/tree/process-exited-errors-5151-2|

The function install in _SIGCHLDWaker looks like:
```
    def install(self):
        """
        Install the handler necessary to make this waker active.
        """
        _signals.installHandler(self.o)
```
Here self.o may evaluate to None causing installHandler to throw an exception. This may happen if connectionLost is somehow called before installHandler.

One possible fix is:
```
    def install(self):
        """
        Install the handler necessary to make this waker active.
        """
        if self.o is not None:
            _signals.installHandler(self.o)
```

I've observed this in the wild inside a moderately complex test suite. Assistance in understanding the "somehow" that causes self.o to become None would be appreciated.

The following note inside _handleSignals in posixbase might be relevant since it explains one possible scenario in which related code may be executed before the event handlers are installed:
```
# Also reap all processes right now, in case we missed any
# signals before we installed the SIGCHLD waker/handler.
# This should only happen if someone used spawnProcess
# before calling reactor.run (and the process also exited
# already).
```

Attachments:

* [waker-process.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c71/c71ce8327ddb80f5f2d66ef497a4c2f2ed295a15/b6fc97ff9e4dd9ced17a43cf06cdc7a6a62bd129.patch) (1030 bytes) - added by exarkun on 2011-06-10 22:27:00Z - 
* [test-process-exited-raises.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c71/c71ce8327ddb80f5f2d66ef497a4c2f2ed295a15/8907444bf68572b60f8da7469d1601e23afd1df8.diff) (1197 bytes) - added by hodgestar on 2011-06-11 13:10:00Z - The first draft of a test that spawns a process and explodes the reactor.
* [test-process-exited-raises.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c71/c71ce8327ddb80f5f2d66ef497a4c2f2ed295a15/17d0d931abb70efd8c52242d09a7752d272e95af.diff) (1247 bytes) - added by hodgestar on 2011-06-11 13:39:06Z - Less broken version of the simple test that raises during processExited and explodes the reactor. Fails (and explodes all later tests) without exarkun's patch. Passes with it.
* [test-process-exited-raises.3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c71/c71ce8327ddb80f5f2d66ef497a4c2f2ed295a15/9d0c5b0d86b17cd0ab2004cac8004268958f0f51.diff) (1394 bytes) - added by hodgestar on 2011-06-11 14:12:43Z - Version of previous patch that doesn't wreck the rest of the test suite if the test fails.
* [test-process-exited-raises.4.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c71/c71ce8327ddb80f5f2d66ef497a4c2f2ed295a15/872ae442400655953cef21592693d2f287f61cb9.diff) (1550 bytes) - added by hodgestar on 2011-06-13 09:27:56Z - Even better test that explicitly raises an error on failure and doesn't erroneously fail if the processExited exception is logged.
* [issue-5151-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c71/c71ce8327ddb80f5f2d66ef497a4c2f2ed295a15/2740e3caf2e21be262656177a30253dd9fab8f84.diff) (2687 bytes) - added by hodgestar on 2011-06-14 13:44:49Z - Single patch with fix, test and NEWS update.

<details><summary>Searchable metadata</summary>

```
trac-id__5151 5151
type__defect defect
reporter__hodgestar hodgestar
priority__normal normal
milestone__ 
branch__branches_process_exited_errors_5151_2 branches/process-exited-errors-5151-2
branch_author__itamarst__hodgestar itamarst, hodgestar
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1307707902000000 1307707902000000
changetime__1388956056000000 1388956056000000
version__None None
owner__hodgestar hodgestar
cc__itamarst
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="hodgestar's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|hodgestar commented|
|-|-|

Steps to reproduce:
```
wget http://pypi.python.org/packages/source/k/katcp/katcp-0.3.4.tar.gz
tar xzf katcp-0.3.4.tar.gz 
cd katcp-0.3.4/
sed -i -e 's/30)/35)/' katcp/tx/test/test_core.py # arrange for a test failure
trial katcp/tx/test # watch all sorts of other tests explode in interesting ways
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Part of the problem here is that `_SIGCHLDWaker.doRead` might raise an exception due to its call to `process.reapAllProcesses`.  Normally `reapAllProcesses` should not raise any exceptions, but it appears to be in this case for some reason.

Either `_SIGCHLDWaker` should put a try/except around that call, or `reapAllProcesses` should be more careful not to raise any exceptions.

The exception being raised here is:

```
          File "/home/exarkun/Projects/Twisted/trunk/twisted/internet/posixbase.py", line 575, in _doReadOrWrite
            why = selectable.doRead()
          File "/home/exarkun/Projects/Twisted/trunk/twisted/internet/posixbase.py", line 230, in doRead
            process.reapAllProcesses()
          File "/home/exarkun/Projects/Twisted/trunk/twisted/internet/process.py", line 46, in reapAllProcesses
            process.reapProcess()
          File "/home/exarkun/Projects/Twisted/trunk/twisted/internet/process.py", line 308, in reapProcess
            log.msg('Failed to reap %d:' % self.pid)
        exceptions.TypeError: %d format: a number is required, not NoneType
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

This happens if application code raises an exception from `processExited`.  The transport implementation doesn't deal with this well, and ends up leaving the dead process's reap handler in the dictionary.  Next time a SIGCHLD arrives, the dead process's handler blows up like this and kills the waker.

Attaching a simple patch that probably fixes the issue; lacking unit tests, though.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

As I see it, there are two problems:

  1. If reapAllProcess raises, sigchld handler install blows up. Notice there's another path which can raise exceptions (unregisterReapProcessHandler).
  2. If processExited raises, reapAllProcesses stops. This is different because it might e.g. cause a second process to never be reaped (at least, until reapAllProcesses is called again, which may be never.)

The patch fixes (1). The test is made to pass by fixing (1), but fixing (2) would also make it pass.

The test should be changed to only test (1) (e.g. make reapAllProcesses raise via monkeypatching or something?). (2) should either be fixed (try/except around processExited call) and tested, or should get its own ticket, up to you.

**Comment by twisted-trac** [comment]
|<img alt="hodgestar's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|hodgestar commented|
|-|-|

Replying to [itamar](#note_5):
> As I see it, there are two problems:

The intent of the test is to check whether the reactor is still healthy after processExited raises.  This seems closer to (2) rather than (1).

Regarding (1), is there an example other than (2) in which reapAllProcesses might raise? If there is another way, I suggest that other way be spun out into a separate ticket instead.

Agreed that only the patch to either posixbase.py or _baseprocess.py is needed to make the test pass. I would prefer the patch to _baseprocess.py to go in.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst commented|
|-|-|

As I mentioned, unregisterReapProcessHandler may in theory raise an exception, which will also break the reactor. So the initial reported symptom is (1), not (2).

I still don't think the test is sufficient, because it (I assume) is passing, it is passing with a fix for (1) even though it's intended to be testing a fix for (2).

**Comment by twisted-trac** [comment]
|<img alt="hodgestar's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|hodgestar commented|
|-|-|

Actually my latest test does appear to distinguish between exarkun's two fixes (I should have noticed this earlier). The fix to posixbase no longer makes the test pass. The fix to _baseprocess does. If processEnded raises unregisterReapProcessHandler is never called and the test picks up the un-reaped process. Any fix to only unregisterReapProcessHandler no longer passes the test because the unregisterReapProcessHandler isn't called in the case where the test fails.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [32119]) Branching to 'process-exited-errors-5151'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [32120]) Apply hodgestar's issue-5151-1.diff

refs [#5151](https://github.com/twisted/twisted/issues/5151)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to hodgestar|
|-|-|

Thanks!

  1. Some minor stuff:
    1. Test docstrings should just say what behavior they're testing, without words like "Test that ..." or "should".  Also, the test doesn't actually verify that the reactor doesn't explode, and no one really knows what "explode" means here anyway, so it's not a good thing to say. :)
    1. You can use `flushLoggedErrors` like this, and you probably should: `self.assertEqual(1, len(self.flushLoggedErrors(TestException)))`.  This both lets your test pass even with the error logged _and_ makes you confident it actually was logged.
    1. The news fragment doesn't go in the `NEWS` file yet. :)  It goes into "twisted/topfiles/5151.bugfix" file.  This is a trick we use to avoid lots of NEWS conflicts.
  1. Some deeper things:
    1. In a `ReactorBuilder`-style test, use `self.runReactor(reactor)` instead of `reactor.run()`.  This gives you a timeout and maybe some other stuff that's nice.
    1. `reactor.callLater(0.01, ...)` is a little gross.  Magic timing numbers mean fragile tests and unreproducable behavior.  `reactor.callLater(0, ...)` is usually okay, because `0` really just means "after I'm done here".  Does that work in this scenario, or is the 0.01 actually doing something?  Another approach might be to stop the reactor in the `processEnded` callback.  And another might even be to just stop it synchronously in `processExited`.
    1. The test needs to skip on Windows, where there is no `process` module (notice how the process module is imported at the top of `test_process.py`.  There are some examples of platform-specific skips in this module already that should demonstrate how to do this.
    1. The inspection of the process module state is also somewhat whitebox-y, in what could otherwise be a nice blackbox test.  I wonder if it could be eliminated (and therefore the need to skip on windows eliminated, too) by verifying success a different way - perhaps running a second process successfully?  However, some of the earlier comments on this ticket confuse me, and I haven't read the earlier versions of the patch.  Feel free to tell my why the way the test is written now is the way it should stay written.

(suddenly out of time, but I think that's most of a review, sorry in advance if it turns out I missed anything :)

I checked your patch and some minor fixes to it into the branch now linked on the ticket.  Please make future patches against the branch.  Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [32122]) minor fixes

refs [#5151](https://github.com/twisted/twisted/issues/5151)

**Comment by twisted-trac** [comment]
|<img alt="hodgestar's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_16"></a>|hodgestar set status to `assigned`|
|-|-|

Replying to [exarkun](#note_14):

> Test docstrings should just say what behavior they're testing, without words like "Test that ..." or "should".  Also, the test doesn't actually verify that the reactor doesn't explode, and no one really knows what "explode" means here anyway, so it's not a good thing to say. :)

Here "explode" means that global state the reactor relies on (i.e. the registered process reap handlers) becomes inconsistent and the reactor is no longer usable to spawn new processes or proceed with other tests. The whitebox-y check at the end of the test cleans up any broken state so that later tests can proceed.

> You can use `flushLoggedErrors` like this, and you probably should: `self.assertEqual(1, len(self.flushLoggedErrors(TestException)))`.  This both lets your test pass even with the error logged _and_ makes you confident it actually was logged.

Cool (I learned something new) and thanks for fixing this in the branch.

> The news fragment doesn't go in the `NEWS` file yet. :)  It goes into "twisted/topfiles/5151.bugfix" file.  This is a trick we use to avoid lots of NEWS conflicts.

Cool (and thanks for fixing this too).

> In a `ReactorBuilder`-style test, use `self.runReactor(reactor)` instead of `reactor.run()`.  This gives you a timeout and maybe some other stuff that's nice.

Will fix.

>  `reactor.callLater(0.01, ...)` is a little gross.  Magic timing numbers mean fragile tests and unreproducable behavior.  `reactor.callLater(0, ...)` is usually okay, because `0` really just means "after I'm done here".  Does that work in this scenario, or is the 0.01 actually doing something?  Another approach might be to stop the reactor in the `processEnded` callback.  And another might even be to just stop it synchronously in `processExited`.

Using 0 here is fine. Just reactor.stop() works too. Will fix. The reactor.callLater(0.01, ...) was the result of a variety of experiments to try regain control in the case were the test fails.

> The test needs to skip on Windows, where there is no `process` module (notice how the process module is imported at the top of `test_process.py`.  There are some examples of platform-specific skips in this module already that should demonstrate how to do this.

> The inspection of the process module state is also somewhat whitebox-y, in what could otherwise be a nice blackbox test.  I wonder if it could be eliminated (and therefore the need to skip on windows eliminated, too) by verifying success a different way - perhaps running a second process successfully?  However, some of the earlier comments on this ticket confuse me, and I haven't read the earlier versions of the patch.  Feel free to tell my why the way the test is written now is the way it should stay written.

See comment replying to the first question -- something needs to fix reapProcessHandlers if the test fails. I'm not sure if similar clean up needs to happen on Windows (I don't have Windows). If not, we could just skip this part of the test if process is not imported (since now the logging of the TestException is checked).

Thanks for the review.

**Comment by twisted-trac** [comment]
|<img alt="hodgestar's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|hodgestar commented|
|-|-|

(In [32127]) Use self.runReactor instead of calling reactor.run directly.

refs [#5151](https://github.com/twisted/twisted/issues/5151)

**Comment by twisted-trac** [comment]
|<img alt="hodgestar's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_18"></a>|hodgestar commented|
|-|-|

(In [32128]) Call reactor.stop directly rather than via an unnecessary reactor.callLater.

refs [#5151](https://github.com/twisted/twisted/issues/5151)

**Comment by twisted-trac** [comment]
|<img alt="hodgestar's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_19"></a>|hodgestar commented|
|-|-|

(In [32129]) Perform process module read handler clean-up only when the process module
is available. Move assertion that process exited exception was logged until
after the clean-up is finished.

refs [#5151](https://github.com/twisted/twisted/issues/5151)

**Comment by twisted-trac** [comment]
|<img alt="hodgestar's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_20"></a>|hodgestar commented|
|-|-|

All the comments from the previous review have been dealt with in the branch now so I think it's ready for the next one.

**Comment by twisted-trac** [comment]
|<img alt="hodgestar's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|hodgestar removed owner<br>hodgestar set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun set owner to hodgestar|
|-|-|

Thanks!  I think the branch is in good shape to be merged.

However, it would be nice if `test_processExitedRaises` could eventually stop poking the `process` module.  I think this means we should have some direct unit tests for `twisted.internet.process` (perhaps in `twisted/internet/test/test_posixprocess.py`) for various exceptional cases of `reapAllProcesses`, `registerReapProcessHandler`, and `unregisterReapProcessHandler`.  These can be as whitebox as necessary, since `twisted.internet.process` is just a big implementation detail.  The other cases where an exception might be raised (mentioned by Itamar above) are interesting too, and we want to make sure they're also being handled correctly.

One other trivial thing: the news file puts "Broken process handlers" up front, but I think people reading the news file will be more interested in "a ProcessProtocol.processExited that raises an exception".  Can you reformulate the news fragment to change the focus so it's more understandable to an application developer who might be affected by the change?

If you do that and file a new ticket the first point and mention it in the comment above the `process`-inspecting part of `test_processExitedRaises` (and if the [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/process-exited-errors-5151) look good), then I think the branch will be in great shape to be merged.

Thanks again.

**Comment by twisted-trac** [comment]
|<img alt="hodgestar's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_23"></a>|hodgestar commented|
|-|-|

(In [32204]) Re-phrase NEWS file entry so that the root cause from an application
developer's perspective is more prominent.

refs [#5151](https://github.com/twisted/twisted/issues/5151)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_24"></a>|@itamarst commented|
|-|-|

Looks like all that needs doing is filing a new ticket, mentioning in a test, and then this can be merged?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_25"></a>|@itamarst commented|
|-|-|

Filed the new ticket: [#6889](https://github.com/twisted/twisted/issues/6889)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_26"></a>|@itamarst commented|
|-|-|

(In [41114]) Branching to 'process-exited-errors-5151-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_27"></a>|@itamarst set status to `closed`|
|-|-|

(In [41116]) Merge process-exited-errors-5151-2.

Fixes: [#5151](https://github.com/twisted/twisted/issues/5151)
Author: hodgestar
Reviewer: exarkun


Handle exceptions raised by ProcessProtocol.processExited.
